native/ios/WatermelonDB/Database.cpp

```diff
@@ -1 +1,533 @@
+                jsi::Array arguments = operation.getValueAtIndex(rt, 4).getObject(rt).getArray(rt);
+
+                executeUpdate(sql, arguments);
+                addedIds.push_back(cacheKey(table.utf8(rt), id));
+            } else if (type == "execute") {
+                jsi::String sql = operation.getValueAtIndex(rt, 2).getString(rt);
+                jsi::Array arguments = operation.getValueAtIndex(rt, 3).getObject(rt).getArray(rt);
+
+                executeUpdate(sql.utf8(rt), arguments);
+            } else if (type == "markAsDeleted") {
+                const jsi::String id = operation.getValueAtIndex(rt, 2).getString(rt);
+                auto args = jsi::Array::createWithElements(rt, id);
+                executeUpdate("update " + table.utf8(rt) + " set _status='deleted' where id == ?", args);
+
+                removedIds.push_back(cacheKey(table.utf8(rt), id.utf8(rt)));
+            } else if (type == "destroyPermanently") {
+                const jsi::String id = operation.getValueAtIndex(rt, 2).getString(rt);
+                auto args = jsi::Array::createWithElements(rt, id);
+
+                // TODO: What's the behavior if nothing got deleted?
+                executeUpdate("delete from " + table.utf8(rt) + " where id == ?", args);
+                removedIds.push_back(cacheKey(table.utf8(rt), id.utf8(rt)));
+            } else {
+                throw jsi::JSError(rt, "Invalid operation type");
+            }
+        }
+        commit();
+    } catch (const std::exception &ex) {
+        rollback();
+        throw;
+    }
+
+    for (auto const &key : addedIds) {
+        markAsCached(key);
+    }
+
+    for (auto const &key : removedIds) {
+        removeFromCache(key);
+    }
+}
+
+jsi::Array Database::getDeletedRecords(jsi::String &tableName) {
+    auto &rt = getRt();
+    auto args = jsi::Array::createWithElements(rt);
+    auto statement = executeQuery("select id from " + tableName.utf8(rt) + " where _status='deleted'", args);
+
+    jsi::Array records(rt, 0);
+
+    for (size_t i = 0; true; i++) {
+        int stepResult = sqlite3_step(statement.stmt);
+
+        if (stepResult == SQLITE_DONE) {
+            break;
+        } else if (stepResult != SQLITE_ROW) {
+            throw dbError("Failed to get deleted records");
+        }
+
+        assert(sqlite3_data_count(statement.stmt) == 1);
+
+        const char *idText = (const char *)sqlite3_column_text(statement.stmt, 0);
+        if (!idText) {
+            throw jsi::JSError(rt, "Failed to get ID of a record");
+        }
+
+        jsi::String id = jsi::String::createFromAscii(rt, idText);
+        records.setValueAtIndex(rt, i, id);
+    }
+
+    return records;
+}
+
+void Database::destroyDeletedRecords(jsi::String &tableName, jsi::Array &recordIds) {
+    auto &rt = getRt();
+    beginTransaction();
+    try {
+        // TODO: Maybe it's faster & easier to do it in one query?
```

Comment:
You could gather IDs into list and perform `delete from (...) where id IN (...IDs)`

Reply:
yeah, I know, but i can't pass an array as a placeholder argument, so I'd have to concatenate strings which I wanted to avoid doing in c++. I'm gonna leave it like that for now, but i'll get back to deletion performance later and see if this matters